Performance optimization (model inheritance)

skulonen opened the pull request:
# Description

**What?**

This pull request ensures that instances of inherited models (i.e. models derived from ModelWithInheritance) are (almost*) always fetched as their actual subclasses, whether it be from a standard QuerySet or a relationship (many-to-many, one-to-one, one-to-many; forwards or backwards).

*) If a related object is selected using `select_related`, e.g. `Submission.objects.select_related('exercise')`, the exercises are not returned as their subclasses.

**Why?**

The `as_leaf_class` method has been used for this purpose, but it has terrible performance characteristics, especially in loops, because every call causes a new database query. This PR removes the need to call it in most cases and is therefore a performance optimization.

**How?**

Most of the work is done by `InheritanceManager` from the `django-model-utils` library. All models derived from ModelWithInheritance now use this manager as their default manager, and the `select_subclasses` call in ModelWithInheritanceManager ensures that inherited models are always fetched as their subclasses when queried normally.

`InheritanceManager` has one limitation: **single related objects**, e.g. `Submission.exercise` are not fetched as their subclasses. This is because internally in Django, `ForeignKey` and `OneToOneField` use the models **base manager** to fetch the related object. We could set the base manager to be ModelWithInheritanceManager, but that would cause issues with object deletion (see previous PR #969). Therefore this PR also introduces new custom relationship fields that ensure the related object is fetched using the model's **default manager** instead of **base manager**.

All unnecessary `as_leaf_class` calls have been removed.


# Testing

**Remember to add or update unit tests for new features and changes.**

* How to [test your changes in A-plus](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations)
* How to [test accessibility](https://wiki.aalto.fi/display/EDIT/How+to+check+the+accessibility+of+pull+requests)


**What type of test did you run?**

- [ ] Accessibility test using the [WAVE](https://wave.webaim.org/extension/) extension.
- [x] Django unit tests.
- [x] Selenium tests.
- [ ] Other test. *(Add a description below)*
- [ ] Manual testing.

Tested manually that the number of queries has been radically reduced in views like "edit course contents".

**Did you test the changes in**

- [ ] Chrome
- [x] Firefox
- [ ] This pull request cannot be tested in the browser.

**Think of what is affected by these changes and could become broken**

# Translation

- [ ] Did you modify or add new strings in the user interface? ([Read about how to create translation](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations))

# Programming style

- [x] Did you follow our [style guides](https://apluslms.github.io/contribute/styleguides/)?

# Have you updated the README or other relevant documentation?

- [ ] documents inside the doc directory.
- [ ] README.md.
- [ ] Aplus Manual.
- [ ] Other documentation (mention below which documentation).

# Is it Done?

- [ ] Reviewer has finished the code review
- [ ] After the review, the developer has made changes accordingly
- [ ] Customer/Teacher has accepted the implementation of the feature

*Clean up your git commit history before submitting the pull request!*

Mankro commented:
This sounds great! It is not a problem to add new migrations assuming that the production database doesn't explode with the new DefaultForeignKeys. Are there any other drawbacks or disadvantages with the second commit (DefaultForeignKey)? If there aren't any problems, then all of this would be good to have. That is to say, both commits are good.

> After we've decided which approach to take (i.e. whether to include commit 2), I'll need to check which `as_leaf_class` calls are now useless (unless we decide to leave them there for now).

You can remove useless calls if you have the time. 

It is good to keep the commits separate in the final version too so that we also have a version with `ModelWithInheritanceManager` and `select_subclasses()` without the DefaultForeignKeys. If we need to get rid of DefaultForeignKeys some day, then it is easier to see those changes in a commit/commits that don't include other unrelated changes.

Mankro commented on exercise/api/csv/views.py:
```diff
@@ -101,7 +101,7 @@ def list(
             exercise_id__in=ids,
             submitters__in=profiles
-        ).prefetch_related(Prefetch('exercise', BaseExercise.objects.all()), 'notifications', 'files')
+        ).prefetch_related('exercise', 'notifications', 'files')
```
It is not obvious to me why the Prefetch is changed, but great that the shorter form works.

skulonen replied:
There are a couple of similar changes in this PR. `Submission.exercise` is a ForeignKey field. Before commit 2, it would always use the base manager, even in prefetches. `BaseExercise.objects.all()` was used to specify that we want to use the default manager here. In commit 2, the field has been changed to use DefaultForeignKey, which uses the default manager while prefetching, so the longer form is no longer needed.

Mankro reviewed (commented):
I took a quick look at the code and it looks good.

skulonen commented on external_services/views.py:
```diff
@@ -104,7 +105,7 @@ class ListMenuItemsView(CourseInstanceBaseView):
     def get_common_objects(self):
         super().get_common_objects()
-        self.menu_items = self.instance.ext_services.all()
+        self.menu_items = self.instance.ext_services.prefetch_related('service')
```
I think that one should've been in the previous PR, but it got moved into the new branch when I split the PR.

skulonen commented on course/views.py:
```diff
@@ -35,6 +36,10 @@ class HomeView(UserProfileView):
     access_mode = ACCESS.ANONYMOUS
     template_name = "course/index.html"
 
+    def get(self, request, *args, **kwargs):
+        raise Exception([i.service.as_leaf_class() for i in MenuItem.objects.all()])
+        return super().get(request, *args, **kwargs)
+
```
Oops, yeah :D I used that for testing random stuff. I'll remove it.

skulonen commented:
> Are there any other drawbacks or disadvantages with the second commit (DefaultForeignKey)?

Possibly a small performance impact, when accessing related objects and using the default manager would not be necessary. For example, when accessing `Submission.exercise`, we don't necessarily need the exercise as its subclass, or its parent. But when using the default manager, those are fetched by default.

> You can remove useless calls if you have the time.

Got it. I think (almost?) all of them should be useless now, since direct queries, related accesses and prefetches now return subclass instances. I can only come up with one situation where an object is not fetched as its subclass, and that is `select_related`. If we do `Submission.objects.select_related('exercise')`, the related exercise is always a BaseExercise. But anyway, I'll chech the `as_leaf_class` calls.

skulonen commented:
I've removed all unnecessary `as_leaf_class` calls, which was all of them except the ones in model methods that reference related inherited models. I didn't feel comfortable removing those because in model methods we can't really be sure how the current instance (`self`) is fetched. It could be that a Submission instance was fetched with a queryset that looks like `Submission.objects.select_related('exercise')`, and as explained earlier, `select_related` doesn't fetch the related model instances as their subclasses.

So, `as_leaf_class` calls still exist in the following files:

apps/models
exercise/exercise_models
exercise/submission_models
external_services/models

I also updated the PR description and marked it as ready for review.

Mankro commented on edit_course/operations/configure.py:
```diff
@@ -643,14 +641,13 @@ def configure_content(instance: CourseInstance, url: str) -> Tuple[bool, List[st
-                        not isinstance(exercise, BaseExercise)
-                        or exercise.submissions.count() == 0
+                        not isinstance(lobject, BaseExercise)
+                        or lobject.submissions.count() == 0
```
Could you check if the suggested change is OK and we can add it to the master branch afterwards?

skulonen replied:
Yes, that seems to work, and is probably a bit faster.

skulonen replied:
There are also several other places in the codebase where `.count() == 0` or `.count() > 0` is used.

Mankro replied:
well, if you could then make a small pull request for those, that would be nice :smile:

skulonen replied:
New PR for this: https://github.com/apluslms/a-plus/pull/978

skulonen commented:
New PR for documentation: https://github.com/apluslms/a-plus/pull/977
